Source/ARTHTTPPaginatedResponse.m

```diff
@@ -0,0 +1,139 @@
+        callback = ^(ARTHTTPPaginatedResponse *_Nullable result, ARTErrorInfo *_Nullable error) {
+            ART_EXITING_ABLY_CODE(self.rest);
+            dispatch_async(self.userQueue, ^{
+                userCallback(result, error);
+            });
+        };
+    }
+
+    if (!self.relNext) {
+        // If there is no next page, we can't make a request, so we answer the callback
+        // with a nil PaginatedResult. That's why the callback has the result as nullable
+        // anyway. (That, and that it can fail.)
+        callback(nil, nil);
+        return;
+    }
+
+    [self.class executePaginated:self.rest withRequest:self.relNext callback:callback];
+} ART_TRY_OR_REPORT_CRASH_END
+}
+
++ (void)executePaginated:(ARTRest *)rest withRequest:(NSMutableURLRequest *)request callback:(void (^)(ARTHTTPPaginatedResponse *_Nullable result, ARTErrorInfo *_Nullable error))callback {
+ART_TRY_OR_REPORT_CRASH_START(rest) {
+    [rest.logger debug:__FILE__ line:__LINE__ message:@"HTTP Paginated request: %@", request];
+
+    [rest executeRequest:request withAuthOption:ARTAuthenticationOn completion:^(NSHTTPURLResponse *response, NSData *data, NSError *error) {
+        if (response.allHeaderFields[ARTHttpHeaderFieldErrorCodeKey] == nil && error) {
+            callback(nil, [ARTErrorInfo createFromNSError:error]);
+        }
+        else {
+            [[rest logger] debug:__FILE__ line:__LINE__ message:@"HTTP Paginated response: %@", response];
+            [[rest logger] debug:__FILE__ line:__LINE__ message:@"HTTP Paginated response data: %@", [[NSString alloc] initWithData:data encoding:NSUTF8StringEncoding]];
+
+            NSError *decodeError = nil;
+
+            ARTPaginatedResultResponseProcessor responseProcessor = ^(NSHTTPURLResponse *response, NSData *data, NSError **errorPtr) {
+                id<ARTEncoder> encoder = [rest.encoders objectForKey:response.MIMEType];
+                return [encoder decodeToArray:data error:errorPtr];
+            };
+            NSArray *items = error ? nil : responseProcessor(response, data, &decodeError);
+
+            if (decodeError) {
+                callback(nil, [ARTErrorInfo createFromNSError:decodeError]);
+                return;
+            }
+
+            NSDictionary *links = [response extractLinks];
+
+            NSMutableURLRequest *firstRel = [NSMutableURLRequest requestWithPath:links[@"first"] relativeTo:request];
+            NSMutableURLRequest *currentRel = [NSMutableURLRequest requestWithPath:links[@"current"] relativeTo:request];
+            NSMutableURLRequest *nextRel = [NSMutableURLRequest requestWithPath:links[@"next"] relativeTo:request];
+
+            ARTHTTPPaginatedResponse *result = [[ARTHTTPPaginatedResponse alloc] initWithResponse:response items:items rest:rest relFirst:firstRel relCurrent:currentRel relNext:nextRel responseProcessor:responseProcessor];
+
+            callback(result, [ARTErrorInfo createFromNSError:error]);
```

Comment:
I don't understand why the error is passed along here. Shouldn't it be nil?

Reply:
I changed it 861f32b. Thanks